In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Q1. Pandas version

In [2]:
print(f'Pandas Version: {pd.__version__}')

Pandas Version: 2.3.0


In [3]:
df = pd.read_csv(r'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [4]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

# Q2. Records count
* How many records are in the dataset?

In [5]:
df.shape

(9704, 11)

we have `9704` records in the dataset

# Q3. Fuel types
* How many fuel types are presented in the dataset?

In [6]:
print('We have {} fuel types in the dataset'.format(df['fuel_type'].nunique()))

We have 2 fuel types in the dataset


# Q4. Missing values
* How many columns in the dataset have missing values?

In [7]:
(df.isnull().sum() > 0).sum()

np.int64(4)

We have 4 columns with null values

# Q5. Max fuel efficiency
* What's the maximum fuel efficiency of cars from Asia?

In [9]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(23.759122836520497)

# Q6. Median value of horsepower
1. Find the median value of `horsepower` column in the dataset.
2. Next, calculate the most frequent value of the same `horsepower` column.
3. Use `fillna` method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of `horsepower` once again.

Has it changed?
* **`Yes, it increased`**
* Yes, it decreased
* No

In [10]:
df['horsepower'].median()

np.float64(149.0)

In [11]:
df["horsepower"].mode()[0]

np.float64(152.0)

In [12]:
df['horsepower'] = df['horsepower'].fillna(df["horsepower"].mode()[0])

In [13]:
df['horsepower'].median()

np.float64(152.0)

# Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?
> **Note:** You just implemented linear regression. We'll talk about it in the next lesson.

* 0.051
* 0.51
* 5.1
* 51

In [14]:
# 1. Select all the cars from Asia
df2 = df[df['origin'] == 'Asia']

In [15]:
# 2. Select only columns vehicle_weight and model_year
df2[['vehicle_weight', 'model_year']]

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [16]:
# 3. Select the first 7 values
df2[['vehicle_weight', 'model_year']].head(7)

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [17]:
# 4. Get the underlying NumPy array. Let's call it X.
X = df2[['vehicle_weight', 'model_year']].head(7).to_numpy()

In [18]:
# 5. Compute matrix-matrix multiplication between the transpose of X and X.
# To get the transpose, use X.T. Let's call the result XTX.
XTX = X.T @ X

In [19]:
# 6. Invert XTX.
XTX_inv = np.linalg.inv(XTX)

In [20]:
# 7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [21]:
# 8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = XTX_inv @ X.T @ y

In [22]:
# 9. What's the sum of all the elements of the result?
print(f'Sum of elements in w: {w.sum()}')

Sum of elements in w: 0.5187709081074016
